src/main/java/com/snowflake/kafka/connector/internal/streaming/StreamingClientProvider.java

```diff
@@ -0,0 +1,208 @@
+    this.streamingIngestClient = client;
+  }
+
+  /**
+   * Creates a new streaming client, will replace the existing one if necessary
+   *
+   * @param connectorConfig The connector config to define the client
+   */
+  public void createClient(Map<String, String> connectorConfig) {
+    // replace previous connector config and client if applicable
+    if (!this.connectorConfig.isEmpty()) {
+      LOGGER.warn("Overriding previous connector config");
+    }
+
+    if (isClientValid(this.streamingIngestClient)) {
+      LOGGER.warn(
+          "Replacing previous valid streaming client. ClientName:{}",
+          this.streamingIngestClient.getName());
+      this.closeClient();
+    }
+
+    this.connectorConfig = connectorConfig;
+    this.streamingIngestClient = this.initStreamingClient(this.connectorConfig);
+  }
+
+  /** Closes the current client */
+  public void closeClient() {
+    // don't do anything if client is already invalid
+    if (!isClientValid(this.streamingIngestClient)) {
+      return;
+    }
+
+    LOGGER.info("Closing Streaming Client:{}", this.streamingIngestClient.getName());
+    try {
+      this.streamingIngestClient.close();
+    } catch (Exception e) {
+      // the client should auto close, so don't throw an exception here
+      String message =
+          e.getMessage() == null || e.getMessage().isEmpty()
+              ? "missing exception message"
+              : e.getMessage();
+      String cause =
+          e.getCause() == null || e.getCause().getStackTrace() == null
+              ? "missing exception cause"
+              : Arrays.toString(e.getCause().getStackTrace());
+
+      LOGGER.error("Failure closing Streaming client msg:{}, cause:{}", message, cause);
+    }
+  }
+
+  /**
+   * Gets the current client or creates a new one from the given connector config If client
+   * optimization is not enabled, just create a new streaming client
+   *
+   * @param connectorConfig The connector config, given as a backup in case the current client is
+   *     invalid
+   * @return The current or newly created streaming client
+   */
+  public SnowflakeStreamingIngestClient getClient(Map<String, String> connectorConfig) {
+    if (Boolean.parseBoolean(
+        connectorConfig.get(
+            SnowflakeSinkConnectorConfig.ENABLE_STREAMING_CLIENT_OPTIMIZATION_CONFIG))) {
+      // recreate streaming client if needed
+      if (!isClientValid(this.streamingIngestClient)) {
+        LOGGER.error("Current streaming client is invalid, recreating client");
+        this.createClient(connectorConfig);
+      }
+      return this.streamingIngestClient;
+    } else {
+      return this.initStreamingClient(connectorConfig);
+    }
+  }
+
+  /**
+   * Initialize the streaming client
+   *
+   * @param connectorConfig The connector config required to create the client
+   * @return An initialized client
+   */
+  private SnowflakeStreamingIngestClient initStreamingClient(Map<String, String> connectorConfig) {
```

Comment:
This is basically the existing behavior copy-pasted from: https://github.com/snowflakedb/snowflake-kafka-connector/blob/1d12734da8b15c0c4ba846b86e902338a68f8827/src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java#L472

Reply:
So the code in SnowflakeSinkServiceV2 will be removed later?

Reply:
yup, that will come in the next PR